Frontc: init at 3.4

maurer opened the pull request:
###### Motivation for this change

Add Frontc OCaml library
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [x] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"` N/A (new package)
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

maurer committed:
Frontc: init at 3.4

rasendubi commented on pkgs/development/ocaml-modules/frontc/default.nix:
```diff
@@ -0,0 +1,37 @@
+{stdenv, buildOcaml, fetchurl}:
```
Use `lib` instead of `stdenv`

rasendubi commented on pkgs/development/ocaml-modules/frontc/default.nix:
```diff
@@ -0,0 +1,37 @@
+    sha256 = "1flhvwr01crn7d094kby0418s1m4198np85ymjp3b4maz0n7m2mx";
+  };
+
+  opam_patch = fetchurl {
```
Is it used?

maurer replied:
Sorry, that was left over from when I was trying to fix the META file issue.

maurer committed:
frontc: trim down unneeded components

maurer commented:
@rasendubi I think my new commit should address your concerns

LnL7 commented on pkgs/development/ocaml-modules/frontc/default.nix:
```diff
@@ -0,0 +1,32 @@
+  postInstall  = ''
+     cp ${meta_file} META
+     ocamlfind install FrontC $OCAMLFIND_DESTDIR/frontc/* META
+     rm -rf $OCAMLFIND_DESTDIR/frontc
```
The output only contains binaries, why is this removed?

maurer replied:
If you remove this line, you'll end up with both a /frontc and a /FrontC directory. The few packages which depend on this (not in nixpkgs atm but exist in opam) refer to FrontC, not frontc.

The patch that was previously referenced as `opam_patch` was the patch opam used to accomplish a similar renaming by editing the makefile, but this ended up being easier.

LnL7 replied:
I'm not sure if it's darwin specific or filesystem related, but on my darwin machine only has `frontc` in the output before the postInstall hook. This means that you end up with an empty `lib/ocaml/4.01.0/site-lib`

maurer replied:
Hm, this sounds like a case sensitivity issue. I'll see if I can get the opam approach to work instead (where they patch the install script).

(Basically, OPAM, and packages depending on this, refer to FrontC, but it installs itself as frontc by default. IIRC HFS+ is case insensitive by default, so on OSX, my hack to move everything to FrontC does not work correctly).

LnL7 replied:
that looks better:
```
darwin> ls -l /nix/store/qpdm4z5kgk4rj375cl765nympc0a0689-ocaml-FrontC-3.4/lib/ocaml/4.01.0/site-lib
total 17
dr-xr-xr-x  13 lnl  staff  13 Jan  1  1970 FrontC

nixos> ls -l /nix/store/q7v8nxhhcd8g0wgyjcfwjb7l56xaxq34-ocaml-FrontC-3.4/lib/ocaml/4.01.0/site-lib                                                                                                                ~/.nix-defexpr/nixpkgs
total 9
dr-xr-xr-x 2 root root 13 Jan  1  1970 FrontC
```

maurer committed:
FrontC: Fix case sensitivity issue

On OSX, the installation method I used previously would result in an empty
package due to frontc and FrontC being treated as the same. However, linux
still requires me to perform that step specifically because frontc and
FrontC are not the same.

Instead, we now patch the Makefile to ensure it ends up in FrontC to begin
with.

rasendubi reviewed (changes requested):
Seems that you have reverted changes I've requested before.

maurer committed:
FrontC: stdenv -> lib

maurer commented:
Sorry, accidental reversion of the stdenv/lib change when I was bringing back the `opam_patch` (which is actually used now.

maurer commented:
Sorry, accidental reversion of the stdenv/lib change when I was bringing back the opam_patch (which is actually used now. @rasendubi
